apps/server/src/api-data/db/migration/db.migration.v3.ts

```diff
@@ -0,0 +1,410 @@
+ * - in case of a newer v3 project we can just return Automation settings
+ * - recover older osc and http subscriptions
+ */
+export function migrateAutomations(jsonData: object): AutomationSettings | undefined {
+  if (is.objectWithKeys(jsonData, ['automation']) && is.object(jsonData.automation)) {
+    // For now the automation type used i v3 and in v4 are the same but we will need to update this if it changes at some point
+    const oldAutomationSettings = structuredClone(jsonData.automation) as AutomationSettings;
+    return oldAutomationSettings;
+  }
+
+  let foundOldSetting = false;
+  const migratedOldStuff = structuredClone(dbModel.automation);
+  const migratedAutomations: NormalisedAutomation = {};
+  const migratedTriggers: Trigger[] = [];
+
+  if (is.objectWithKeys(jsonData, ['osc']) && is.object(jsonData.osc)) {
+    foundOldSetting = true;
+    const { subscriptions, portIn, enabledIn, targetIP, portOut } = structuredClone(jsonData.osc) as old_OSCSettings;
+    migratedOldStuff.enabledOscIn = enabledIn;
+    migratedOldStuff.oscPortIn = portIn;
+
+    for (const subscription of subscriptions) {
+      const { id, cycle, address, payload } = subscription;
+
+      migratedTriggers.push({
+        id: `${id}-T`,
+        automationId: `${id}-A`,
+        title: `Migrated Trigger ${id}`,
+        trigger: cycle,
+      });
+
+      migratedAutomations[`${id}-A`] = {
+        id: `${id}-A`,
+        title: `Migrated Automation ${id}`,
+        filterRule: 'any',
+        filters: [],
+        outputs: [{ type: 'osc', address, args: payload, targetIP, targetPort: portOut }],
+      };
+    }
+  }
+
+  if (is.objectWithKeys(jsonData, ['http']) && is.object(jsonData.http)) {
+    foundOldSetting = true;
+    const { subscriptions } = structuredClone(jsonData.http) as old_HttpSettings;
+
+    for (const subscription of subscriptions) {
+      const { id, cycle, message } = subscription;
+      migratedTriggers.push({
+        id: `${id}-T`,
+        automationId: `${id}-A`,
+        title: `Migrated Trigger ${id}`,
+        trigger: cycle,
+      });
+
+      migratedAutomations[`${id}-A`] = {
+        id: `${id}-A`,
+        title: `Migrated Automation ${id}`,
+        filterRule: 'any',
+        filters: [],
+        outputs: [{ type: 'http', url: message }],
+      };
+    }
+  }
+
+  if (foundOldSetting) {
+    migratedOldStuff.automations = migratedAutomations;
+    migratedOldStuff.triggers = migratedTriggers;
+    return migratedOldStuff;
+  }
+}
+
+/**
+ * migrates a rundown from v3 to v4.0.0
+ * -  name the rundown default and place it in the multi rundown object
+ * - generate rundown info placeholders (can be somewhat empty as it will be regenerated by the rundown init)
+ *
+ * - events:
+ *  - add flag
+ *  - ensure end action is not stop
+ *  - move the timer type count-to-end to it owen setting
+ *  - ensure triggers
+ *  - add parent
+ *
+ * - block:
+ *  - add all the new blocks of the block that is now a group
```

Comment:
question: when we see a block, could we add all the elements below inside the block until we find a new block?

Reply:
I don’t think there is a clean path that will 100% work. So the decision is whether it is more convenient than having to populate all groups (thinking we don’t have drag and drop for multiple elements)